# Practice 12

### GENERAL STATEMENT

From the ICADEHosp hospital, we have been tasked with conducting a study to diagnose patients at risk of heart attack. To do this, they ask us to:

1.- Create the lists: names, age, obesity, and cholesterol. To do this, the empty lists must be created in the main program and the number of patients (num_pac) must be requested by keyboard input. The minimum number of patients is 3 (therefore **it is necessary to VALIDATE that the number of patients entered is greater than or equal to 3. The request must be repeated until the condition is met**). Then, call the function CreaciónListas(nombres, edad, obesidad, colesterol, num_pac).

    names (string)
    age (int)
    obesity (string) values (YES/NO)
    cholesterol (int)

    Detailed explanation below

2.- Creation of the dictionary dic_enfermos from the previously created lists. To do this, the empty dictionary enfermos must be defined in the main program and call the function CreacionDiccionario(dic_enfermos, nombres, edad, obesidad, colesterol). The structure of the dic_enfermos dictionary is:
key: patient's name
value: list with age, obesity, and cholesterol

    Detailed explanation below

3.- Creation of the enfermos.csv file. To do this, the file must be defined and opened in the main program and call the function CrearFichero(file, dic_enfermos)
the structure of the file will be:
name; age; obesity; cholesterol\n"

    Detailed explanation below

4.- Function CalcularMostrar(file). This function reads the file, calculates the risk of heart attack for each patient, and shows the name of each patient and whether or not they are at risk of heart attack.
The risk indicator is the following:
* If they are over 65 years old and have a very high level of cholesterol

    The names of patients and their risk of heart attack should be displayed on the screen.

    The cholesterol levels are those described in the following table:
|  Cholesterol level   | Category |
|----------------------|----------|
| Below 100            | Optimal  |
| 100 - 129            | Normal   |
| 130 - 159            | High     |
| 160 - and higher     | Very High|


### What has been discussed so far is the context of the exam. We are going to implement this program step by step below. In each section, the requested functions must be completed, and in the final cell, we will have the main function, which will be used to test the calls to these functions.

**Note:** *This program must work for any number of patients and any value that is entered, not only exclusively for those in the example.*


# Section 1

At this point, we will create the **function CreateLists(patients, age, obesity, cholesterol, pat_num)**. This function is responsible for requesting data (Name, age, obesity, cholesterol) for each of the patients (there are num_pac number of patients) and storing them in the respective lists.

Once we have the function coded, we can proceed to **section 5 (main)**: Create the following empty lists: names, age, obesity, and cholesterol, and request the number of patients (pat_num) via keyboard input. The minimum number of patients will be 3 (**therefore it is necessary to VALIDATE that the number of patients entered is greater than or equal to 3. The request must be repeated until the condition is met**). Then call the function CreateLists(names, age, obesity, cholesterol, pat_num).
- `names` (string)
- `age` (int)
- `obesity` (string) values (YES/NO)
- `cholesterol` (int)

In [1]:
def CreateLists(patients, age, obesity, cholesterol, pat_num):
    for i in range(pat_num):
        curr_patient = input(f"Give me the name for patient {i}: ")
        patients.append(curr_patient)
        
        curr_age = int(input(f"Give me the age for patient {i}: "))
        age.append(curr_age)
        
        curr_obesity = input(f"Tell me if patient {i} has obesity: ")
        obesity.append(curr_obesity)
        
        curr_cholesterol = input(f"Give me the cholesterol for patient {i}: ")
        cholesterol.append(curr_cholesterol)
    
    return
 


In [2]:
patients = []
age = []
obesity = []
cholesterol = []

pat_num = int(input('Give me the number of patients: '))
while pat_num <3:
    pat_num = int(input('Give me the number of patients: '))
    
CreateLists(patients, age, obesity, cholesterol, pat_num)

Give me the number of patients: 3
Give me the name for patient 0: Pau
Give me the age for patient 0: 18
Tell me if patient 0 has obesity: YES
Give me the cholesterol for patient 0: 210
Give me the name for patient 1: Lu
Give me the age for patient 1: 32
Tell me if patient 1 has obesity: NO
Give me the cholesterol for patient 1: 300
Give me the name for patient 2: Mario
Give me the age for patient 2: 88
Tell me if patient 2 has obesity: NO
Give me the cholesterol for patient 2: 130


In [3]:
print(patients)
print(age)
print(obesity)
print(cholesterol)

['Pau', 'Lu', 'Mario']
[18, 32, 88]
['YES', 'NO', 'NO']
['210', '300', '130']


# Section 2

At this point, we will create the **function CreateDictionary(sick_dict, names, age, obesity, cholesterol)**. This function creates the dictionary `sick_dict` from the previously created lists. The structure of the `sick_dict` dictionary is:  
key: patient's name  
value: list with age, obesity, and cholesterol

Once we have the function coded, we can proceed to **section 5 (main)**:
We define the empty dictionary `sick` in the main program and call the function CreateDictionary(sick_dict, names, age, obesity, cholesterol).

**NOTE** If you have not solved section 1, manually construct the lists names, age, obesity, and cholesterol with the following values:  
names=["Alvaro","Elena","Manuel","Gonzalo","María","Cristina","Rodrigo"]  
age=[67,45,79,50,82,75,68]  
obesity=["YES","NO","YES","YES","YES","NO","NO"]  
cholesterol=[190,180,150,120,210,180,110]

**The content of the dictionary for these patients would be**  
{'Alvaro': [67, 'YES', 190], 'Elena': [45, 'NO', 180], 'Manuel': [79, 'YES', 150], 'Gonzalo': [50, 'YES', 120], 'María': [82, 'YES', 210], 'Cristina': [75, 'NO', 180], 'Rodrigo': [68, 'NO', 110]}


In [4]:
patients

['Pau', 'Lu', 'Mario']

In [5]:
def CreateDictionary(sick_dict, names, age, obesity, cholesterol):
    for i in range(len(names)):
        sick_dict[names[i]] = [age[i], obesity[i], cholesterol[i]]
    return



In [6]:
sick_dict = {}
CreateDictionary(sick_dict, patients, age, obesity, cholesterol)

In [7]:
sick_dict

{'Pau': [18, 'YES', '210'],
 'Lu': [32, 'NO', '300'],
 'Mario': [88, 'NO', '130']}

# Section 3

At this point, we will create the **function CreateFile(file, sick_dict)**.

This function creates the `sick.csv` file. **Note: the file is opened and closed in the main, not within this function**.
The structure of the file will be:  
1.- A first header line with a title for each column.  
2.- Each line will have the following structure:  
"name;age;obesity;cholesterol\n".  

Once we have the function coded, we can proceed to **section 5 (main)**:
The file must be defined and opened in the main program and call the function CreateFile(file, sick_dict). Afterwards, it must be closed.

An example of the created file would be, for the example, it must be opened again and its content shown, for example:  
Name;Age;Obesity;Cholesterol  
Alvaro;67;YES;190  
Elena;45;NO;180  
Manuel;79;YES;150  
Gonzalo;50;YES;120  
María;82;YES;210  
Cristina;75;NO;180  
Rodrigo;68;NO;110  

**NOTE** If you have not solved section 2, manually construct the dictionary with the following values:  
sick_dict={'Alvaro': [67, 'YES', 190], 'Elena': [45, 'NO', 180], 'Manuel': [79, 'YES', 150], 'Gonzalo': [50, 'YES', 120], 'María': [82, 'YES', 210], 'Cristina': [75, 'NO', 180], 'Rodrigo': [68, 'NO', 110]}  

**The content of the file, after executing with the above data, would be**  
Name;Age;Obesity;Cholesterol  
Alvaro;67;YES;190  
Elena;45;NO;180  
Manuel;79;YES;150  
Gonzalo;50;YES;120  
María;82;YES;210  
Cristina;75;NO;180  
Rodrigo;68;NO;110


In [9]:
sick_dict

{'Pau': [18, 'YES', '210'],
 'Lu': [32, 'NO', '300'],
 'Mario': [88, 'NO', '130']}

In [10]:
def CreateFile(file, sick_dict):
    file.write(f"name;age;obesity;cholesterol\n")
    for name, data in sick_dict.items():
        file.write(f"{name};{data[0]};{data[1]};{data[2]}\n")
    return

In [11]:
my_file = open("patients.txt", "w", encoding="utf8")
CreateFile(my_file, sick_dict)
my_file.close()

# Section 4

At this point, we will create the **function CalculateShow(file)**. This function reads the file, calculates the risk of heart attack for each patient, and displays on the screen the name of each patient and whether they are at risk of heart attack or not.

The risk indicator is "YES" if the patient meets the following criteria:
   * Is over 65 years old and has a very high level of cholesterol

Otherwise, "NO" will be displayed.

Cholesterol levels are described in the following table:

| Cholesterol Level | Category   |
|-------------------|------------|
| Less than 100     | Optimal    |
| 100 - 129         | Normal     |
| 130 - 159         | High       |
| 160 and above     | Very High  |

The names of patients and their heart attack risk should be displayed on the screen, as follows:

**Output of results**  
Name    Risk  
Alvaro      YES  
Elena       NO  
Manuel      NO  
Gonzalo     NO  
María       NO  
Cristina    YES  
Rodrigo     NO

Additionally, this function must calculate the percentage of patients who are at risk. For this, we will keep count of the number of patients in the file in the variable `total_num_pac` and the count of those at risk in `num_pac_risk`. To find the percentage, we will calculate the percentage as `num_pac_risk` multiplied by 100 and divided by `total_num_pac`. This function returns this percentage value.

Once we have the function coded, we can proceed to **section 5 (main)**:
The `sick.csv` file must be opened in the main program and call the function CalculateShow(file). Afterwards, it must be closed.
Finally, a message indicating the percentage of patients at risk should be printed, as shown in the execution example.

In [12]:
def CalculateShow(file):
    file.readline() 
    num_pac_risk = 0
    total_num_pac = 0
    for line in file: 
        line_split = line.strip().split(";") 
        

        total_num_pac = total_num_pac + 1
        print(line_split)
        
        if int(line_split[1]) > 65 and int(line_split[3]) > 160:
            print(f"{line_split[0]}: YES")
            num_pac_risk = num_pac_risk + 1
        else:
            print(f"{line_split[0]}: NO")
            
    perc_patients = num_pac_risk/total_num_pac
    
    return perc_patients
    

In [13]:
file = open("patients.txt", "r", encoding="utf8")
perc_patients = CalculateShow(file)
file.close()
print(round(perc_patients,2))

['Pau', '18', 'YES', '210']
Pau: NO
['Lu', '32', 'NO', '300']
Lu: NO
['Mario', '88', 'NO', '130']
Mario: NO
0.0


# Section 5

Main function that calls the functions from the previous sections. Students are advised to complete it step by step as they have the code from the previous functions. The functionality of the main has been explained in each of the previous sections.

**An example of the output of the main program would be:**

Enter the number of patients: 2    
Minimum 3 patients. Enter the number of patients: 3     
    
    
Enter the patient's name: Alvaro    
Enter the patient's age: 67   
Enter obesity (YES/NO): YES   
Enter cholesterol (0-300): 190     
    
   
Enter the patient's name: Elena    
Enter the patient's age: 45    
Enter obesity (YES/NO): NO    
Enter cholesterol (0-300): 180    
    
    
Enter the patient's name: Manuel   
Enter the patient's age: 79    
Enter obesity (YES/NO): YES    
Enter cholesterol (0-300): 150  
    
The dictionary of sick is:     
{'Alvaro': [67, 'YES', 190], 'Elena': [45, 'NO', 180], 'Manuel': [79, 'YES', 150]}**   


NAME&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;RISK    
$===============$   
Alvaro&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;YES    
Elena&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;NO    
Manuel&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;NO    
    
     
The percentage of patients at risk is 33.33


In [15]:
def main():
    
    #Section 1
    patients = []
    age = []
    obesity = []
    cholesterol = []

    pat_num = int(input('Give me the number of patients: '))
    while pat_num <3:
        pat_num = int(input('Give me the number of patients: '))

    CreateLists(patients, age, obesity, cholesterol, pat_num)
    
    #Section 2
    
    sick_dict = {}
    CreateDictionary(sick_dict, patients, age, obesity, cholesterol)
    
    #Section 3
    
    my_file = open("patients.txt", "w", encoding="utf8")
    CreateFile(my_file, sick_dict)
    my_file.close()
    
    #Section 4
    file = open("patients.txt", "r", encoding="utf8")
    perc_patients = CalculateShow(file)
    file.close()
    print(round(perc_patients,2))

main()



Give me the number of patients: 3
Give me the name for patient 0: Pau
Give me the age for patient 0: 18
Tell me if patient 0 has obesity: No
Give me the cholesterol for patient 0: 200
Give me the name for patient 1: Lu
Give me the age for patient 1: 34
Tell me if patient 1 has obesity: YES
Give me the cholesterol for patient 1: 300
Give me the name for patient 2: Mario
Give me the age for patient 2: 88
Tell me if patient 2 has obesity: NO
Give me the cholesterol for patient 2: 210
['Pau', '18', 'No', '200']
Pau: NO
['Lu', '34', 'YES', '300']
Lu: NO
['Mario', '88', 'NO', '210']
Mario: YES
0.33
